refactor: replace isPotentialtip/waitForNotifications by higher method

ariard opened the pull request:
In Chain interface, instead of a isPotentialTip and a WaitForNotifications method, both used only once in CWallet::BlockUntilSyncedToCurrentChain, combine them in a higher WaitForNotificationsUpToTip method. Semantic should be unchanged, wallet wait for pending notifications to be processed unless block hash points to the current chain tip or a descendant.

Commit:
refactor: replace isPotentialtip/waitForNotifications by higher method

Add GUARDED_BY(cs_wallet) annotation to m_last_block_processed, given
that its now guarded by cs_wallet instead of cs_main

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -1288,24 +1288,11 @@ void CWallet::UpdatedBlockTip()
     // for the queue to drain enough to execute it (indicating we are caught up
     // at least with the time we entered this function).
-    chain().waitForNotifications();
+    chain().waitForNotificationsUpToTip(m_last_block_processed);
```
Accessing `m_last_block_processed` requires the `cs_wallet` lock. I think to make this work you'll need to:
- grab `cs_wallet`
- access `m_last_block_processed`
- release `cs_wallet`
- call the `waitForNotificationsUpToTip` interface method

Alternatively, we could make `m_last_block_processed` an atomic.

Thoughts, @ryanofsky ?

ryanofsky replied:
Yes, John's right this needs to synchronize access to m_last_block_processed. Since I don't think there's a straightforward way to make uint256 atomic, maybe a clean way to do this would be to add a new `sync.h` macro:

```diff
-chain().waitForNotificationsUpToTip(m_last_block_processed);
+chain().waitForNotificationsUpToTip(WITH_LOCK(cs_wallet, return m_last_block_processed));
```

```c++
//! Run code while locking a mutex.
//!
//! Examples:
//!
//!   WITH_LOCK(cs, shared_val = shared_val + 1);
//!
//!   int val = WITH_LOCK(cs, return shared_val);
//!
#define WITH_LOCK(cs, code) [&]{ LOCK(cs); code; }()
```

jnewbery replied:
I like the macro, but I don't like putting the complete expression in the `waitForNotificationsUpToTip()` function call. It's too easy for a casual observer to think that the lock is held for the entire call to `waitForNotificationsUpToTip()` (which of course would cause a deadlock). How about:

```c++
uint256 last_block_hash = WITH_LOCK(cs_wallet, m_last_block_processed);
chain().waitForNotificationsUpToTip(last_block_hash);
```

which perhaps compiles to the same thing.

ariard replied:
Ah yes, wasn't sure of cs_wallet lock covering. Added a WITH_LOCK commit and used to get back m_last_block_processed as showed in example.

ryanofsky commented on src/interfaces/chain.h:
```diff
@@ -271,8 +260,10 @@ class Chain
+    //! Wait for pending notifications to be processed unless block hash points to the current,
+    //! chain tip, or to a possible descendant of the current chain tip that isn't currently
+    //! connected.
+    virtual void waitForNotificationsUpToTip(const uint256& hash) = 0;
```
I think the "wait up to" name I suggested earlier in the todo comment is misleading, because this isn't really deciding what to wait for based on the argument, but deciding whether to wait based on the argument.

I think a better name might be `waitForNotificationsIfNewBlocks(old_tip)` or `waitForNotificationsIfNewBlocksConnected(old_tip)`

ariard replied:
Yes agree wasn't clear so I added `Tip` but even followed `waitForNotificationsIfNewBlocksConnected(old_tip)` suggestion, a little bit more verbose but far more insightful

Commit:
Add WITH_LOCK macro: run code while locking a mutex

Results from ryanofksy suggestion on isPotentialTip/
waitForNotifications refactoring

jnewbery commented on src/interfaces/chain.cpp:
```diff
@@ -358,7 +352,16 @@ class ChainImpl : public Chain
         return MakeUnique<NotificationsHandlerImpl>(*this, notifications);
     }
-    void waitForNotifications() override { SyncWithValidationInterfaceQueue(); }
+    void waitForNotificationsIfNewBlocksConnected(const uint256& hash) override
```
nit: please rename the argument `old_tip` to be consistent with the function declaration.

jnewbery commented on src/interfaces/chain.h:
```diff
@@ -271,8 +260,10 @@ class Chain
 
-    //! Wait for pending notifications to be handled.
-    virtual void waitForNotifications() = 0;
+    //! Wait for pending notifications to be processed unless block hash points to the current,
```
nit: remove trailing comma

jnewbery reviewed (commented):
utACK a4c51ed8ee2b087c1228402c3d4e665b4f0fcf37 with a couple of minor nits.

Please remove the `@ryanofsky` from the commit log. Github buries you in notifications if your name appears in a commit log. You can keep the `ryanofsky` as long as it's not prefixed by `@`.

ariard commented:
Noted for the github trick, solved the nits on deb5f75

jnewbery commented:
Doesn't build. Please build/test locally before pushing :slightly_smiling_face:

ariard commented:
Sorry for that, working on dev scripts to avoid this kind of mistake even for nits updates!

ryanofsky reviewed (approved):
utACK 57007b8c9e6a70065d4c31fad3c81e60f7aeccd9. This change isn't as much of a simplification as I was hoping when I wrote the TODO suggesting it, but it's good to remove this one Chain::Lock usage, just as a prerequisite for removing the entire Chain::Lock interface later.

One more suggestion would be to add `GUARDED_BY(cs_wallet)` annotation to:

https://github.com/bitcoin/bitcoin/blob/cd14d210c4156a0a17450d3d451a92ae3938b019/src/wallet/wallet.h#L690

because after this PR, accesses are now guarded by `cs_wallet` instead of `cs_main`.

ariard commented:
Updated at 4226779 with GUARDED_BY annotation addition.

(btw @ryanofsky, I've started to work on remove bits of Chain::Lock interface, notably the GetDepthInMainChain by tracking last_block_processed_height in CWallet, hope to PR soon to get feedback)

ryanofsky reviewed (approved):
utACK 422677963a7b41e340b911b4cd53d29dd8d63f21. Only change is adding the cs_wallet lock annotation.

I'm glad to hear you're working on removing cs_main locking from GetDepthInMainChain. This should improve performance and allow removing locked_chain variables from a swath of wallet code. I think it'll require adding something like a map from block hash to height for `CMerkleTx::hashBlock` blocks, but there could be other approaches.

ariard commented:
Maybe hold on a bit on this.

While working on removing cs_main locking, it appears to me than, given wallet will be tracking chain height thanks to asynchronous updates and not direct chain state queries, it's needed to ```SyncWithValidationInterfaceQueue``` before each take of cs_wallet lock even if wallet is at the tip, because we may have ```BlockDisconnected``` in the callbacks validation interface queue and so it may cause races between RPC calls (at least spotted one thanks to mempool_resurrect)

And yes doing remove by adding m_last_block_processed_height in both ```CWallet``` and m_block_height in ```CMerkleTx```

ryanofsky commented:
I don't think I understand the need to call `SyncWithValidationInterfaceQueue` unconditionally in the other PR, but even if there's no way around it, this PR doesn't prevent that because it's still possible to call `waitForNotificationsIfNewBlocksConnected({})`

ariard commented:
Ah get your point, I'm going to far in locking removing (at least for now), was trying to not rely at all on Chain locking, even without '''::chainActive.Tip()->GetBlockHash()''' call, but only on callbacks hinting chain state.

jnewbery commented:
ACK 422677963a7b41e340b911b4cd53d29dd8d63f21
